core/src/main/java/org/neo4j/ogm/cypher/compiler/CypherContext.java

```diff
@@ -175,38 +165,43 @@ public boolean deregisterOutgoingRelationships(Long src, String relationshipType
      */
     public boolean deregisterIncomingRelationships(Long tgt, String relationshipType, Class endNodeType,
         boolean relationshipEntity) {
+
         Iterator<Mappable> iterator = registeredRelationships.iterator();
-        List<Mappable> cleared = new ArrayList<>();
-        boolean nothingToDelete = true;
+        List<Mappable> candidatesForDeletion = new ArrayList<>();
         while (iterator.hasNext()) {
             Mappable mappedRelationship = iterator.next();
             if (mappedRelationship.getEndNodeId() == tgt &&
                 mappedRelationship.getRelationshipType().equals(relationshipType) &&
                 endNodeType.equals(
                     relationshipEntity ? mappedRelationship.getEndNodeType() : mappedRelationship.getStartNodeType())) {
 
-                cleared.add(mappedRelationship);
+                candidatesForDeletion.add(mappedRelationship);
                 iterator.remove();
-                nothingToDelete = false;
             }
         }
 
-        if (nothingToDelete) {
-            return true; //relationships not in the graph, okay, we can return
-        }
+        boolean notInGraphAtAll = candidatesForDeletion.size() == 0;
+        return notInGraphAtAll || handleDeletionCandidates(candidatesForDeletion);
+    }
 
-        //Check to see if the relationships were previously deleted, if so, restore them
-        iterator = cleared.iterator();
-        while (iterator.hasNext()) {
-            Mappable mappedRelationship = iterator.next();
-            if (isMappableAlreadyDeleted(mappedRelationship)) {
-                registerRelationship(mappedRelationship);
-                iterator.remove();
-            } else {
-                deletedRelationships.add(mappedRelationship);
-            }
-        }
-        return cleared.size() > 0;
+    /**
+     * Partitions the list of candidates for deletion by relationships that have already been deleted and those that
+     * need to be deleted and than re-register those that already have been marked as deleted or marks all others as deleted.
+     * @param candidatesForDeletion A list of candidates for deletion
+     * @return True, if any candidate needs to be deleted
+     */
+    private boolean handleDeletionCandidates(List<Mappable> candidatesForDeletion) {
+
+        Map<Boolean, List<Mappable>> help = candidatesForDeletion.stream()
```

Comment:
Can you take a look at [my optimization](https://github.com/neo4j/neo4j-ogm/pull/577/files#diff-bc01d57c8046cf702a69422d2376bb40R163), this will reduce list changes to a minimum.